### PR TITLE
create new TA preview API v3.1-preview.2

### PR DESCRIPTION
**Pull request by laramume**
Changes in the PR

- New Preview API v3.1-preview.2 which introduces bingId property in /entities/linking and redactedText in /entities/recognition/pii
- Add a new query param called "stringIndexType" to return offsets based on the value in v3.1-preview.1 API
- Fix Swagger Error Response bug in all V3 API's

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json [hunk 1]
```diff
@@ -60,6 +60,15 @@
...
           },
+          {
+            "name": "stringIndexType",
+            "in": "query",
+            "description": "(Optional) This value indicates the index type for which the offsets have to be returned.",
+            "type": "string",
+            "enum" : [ "textelement_v8", "unicodecodepoint", "utf16codeunit" ],
+            "default" : "textelement_v8",
+            "required": false
+          },
```
Comment (assafi) [inline specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:71]: Please see my v3.2-preview.1 [PR](https://github.com/Azure/azure-rest-api-specs-pr/pull/1728/files). It includes v3.1-preview.1 updates that should be incorporated here as well.

Comment (laramume) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:57]: I have synced this with that PR and this also follows the same conventions now

**Comment by laramume** [comment]
@tjprescott @yangyuan Could you guys take a look at this?

**Comment by tjprescott** [comment]
You will need to reach out to @lmazuel to be able to merge this with breaking changes.

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json [hunk 2]
```diff
@@ -47,27 +47,16 @@
...
-            "name": "input",
-            "description": "Collection of documents to analyze.",
-            "required": true,
-            "schema": {
-              "$ref": "#/definitions/MultiLanguageBatchInput"
-            }
+            "$ref": "#/parameters/MultiLanguageInput"
+          },
+          {
+            "$ref": "#/parameters/StringIndexType"
```
Comment (iscai-msft) [inline specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:59]: is string index type also being added to v3.1-preview.1? My understanding was that it was being added v3.1-preview.2 and onward

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json [hunk 3]
```diff
@@ -77,10 +77,16 @@
               "$ref": "#/definitions/EntitiesResult"
             }
           },
-          "default": {
```
Comment (johanste) [inline specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json:80]: You probably want to keep the default (e.g. what are you returning on a 429 response?)

**Review by johanste — APPROVED** [review]
Please consider keeping the "default" error declaration rather than the individual error codes. If you *do* want individual error codes, they should be marked with [`x-ms-error-response`](http://azure.github.io/autorest/extensions/#x-ms-error-response) to clearly indicate to autorest and friends that it is an error case/expected to raise an exception.

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json [hunk 4]
```diff
@@ -0,0 +1,1481 @@
...
+                 "value":"UnicodeCodePoint",
+                 "description":"Returned offset and length values will correspond to Unicode code points. Use this option if your application is written in a language that support Unicode, for example Python."
+              },
+              {
+                 "value":"Utf16CodeUnit",
+                 "description":"Returned offset and length values will correspond to UTF-16 code units. Use this option if your application is written in a language that support Unicode, for example Java, JavaScript."
+              }
+           ]
+        },
+        "x-ms-parameter-location":"client"
```
Comment (iscai-msft) [inline specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json:1442]: Why are you choosing to make it a client parameter instead of a method parameter?

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json (hunk 2 above)
Comment (laramume) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:59]: Yes its being added to v3.1-preview.1 also. I was not in the discussions for it. @thediris might know better

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json (hunk 4 above)
Comment (iscai-msft) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json:1442]: For python multiapi, it'll be better to have this on the method level. We haven't dealt with client parameters that are required for some versions, and don't exist for others

Comment (laramume) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json:1442]: @assafi had this defined in another PR which I just copied over. Assaf is there a reason to have it like this? @thediris if you have an answer for this please let us know

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json (hunk 3 above)
Comment (laramume) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json:80]: Sure I will add those

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json (hunk 4 above)
Comment (iscai-msft) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json:1442]: @johanste do you have any input regarding string index type being a client parameter?

Comment (johanste) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json:1442]: I believe this was based on the assumption that we wouldn't provide a parameter at all for the encoding. But based on the conversation in the arch board review (scenario where I may want to change it when storing something in a database to be consumed by a different language etc.), it is probably better to (still not exposing it as a parameter in the first release) have it as a per-method parameter.

Comment (laramume) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json]: I have changed it to method. @assafi Please let us know if you have any concerns

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json (hunk 3 above)
Comment (laramume) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json:80]: Added

**Review by tjprescott — APPROVED** [review]

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json [hunk 5]
```diff
@@ -1337,6 +1372,73 @@
...
+            "value": "UnicodeCodePoint",
+            "description": "Returned offset and length values will correspond to Unicode code points. Use this option if your application is written in a language that support Unicode, for example Python."
+          },
+          {
+            "value": "Utf16CodeUnit",
+            "description": "Returned offset and length values will correspond to UTF-16 code units. Use this option if your application is written in a language that support Unicode, for example Java, JavaScript."
+          }
+        ]
+      },
+      "x-ms-parameter-location": "client"
```
Comment (iscai-msft) [inline specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:1405]: looks like the parameter location for this string index type is still client

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json [hunk 6]
```diff
@@ -77,10 +77,22 @@
               "$ref": "#/definitions/EntitiesResult"
             }
           },
```
Comment (assafi) [inline specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json:79]: nit: We will be deprecating v3.0-preview.1 on 9/4 so we can probably get rid of this spec completely.

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json [hunk 7]
```diff
@@ -0,0 +1,1443 @@
...
+          "type": "string",
+          "enum": [
+            "InvalidRequest",
+            "InvalidArgument",
+            "InternalServerError",
+            "ServiceUnavailable"
+          ],
+          "x-ms-enum": {
+            "name": "ErrorCodeValue",
+            "modelAsString": false
```
Comment (assafi) [inline specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json:423]: I think this might need to be `true`

**Review by iscai-msft — APPROVED** [review]

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json (hunk 6 above)
Comment (laramume) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.0-preview.1/TextAnalytics.json]: I will handle this in a separate PR if its not an issue

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json (hunk 7 above)
Comment (laramume) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json]: Changed it

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json (hunk 4 above)
Comment (assafi) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.2/TextAnalytics.json]: Yep. Agreed with everything that's been said. Thanks for the feedback everyone.

**Review by johanste — APPROVED** [review]

File: specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json [hunk 8]
```diff
@@ -849,6 +804,10 @@
             "mixed",
             "negative"
           ],
+          "x-ms-enum": {
+            "name": "TokenSentimentValue",
+            "modelAsString": false
+          },
```
Comment (deyaaeldeen) [inline specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:810]: why was this added? cc @iscai-msft

Comment (iscai-msft) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:810]: This changes the name of the generated enum class, and ensures that the enum is non-extensible, I'm assuming those were the reasons it was added, @laramume for confirmation

Comment (deyaaeldeen) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:810]: So it is an implementation of the new guidelines defined here: https://github.com/microsoft/api-guidelines/pull/213. Thanks @iscai-msft.

Comment (assafi) [reply specification/cognitiveservices/data-plane/TextAnalytics/preview/v3.1-preview.1/TextAnalytics.json:810]: I'm actually not sure myself what this is. @laramume - please remove this if not needed.
